lightcrafts/jnisrc/libraw/LibRawJNI.cpp

```diff
@@ -14,294 +12,297 @@
-    if (libRaw == nullptr)
-        return nullptr;
-
-    int status = libRaw->unpack();
-    if (status != LIBRAW_SUCCESS)
-        return nullptr;
-
-    // params
-    if (half_size)
-        libRaw->imgdata.params.half_size = 1;
-    else if (interpolate)
-        libRaw->imgdata.params.four_color_rgb = 1;
-    else
-        libRaw->set_interpolate_bayer_handler(interpolate_bayer);
-    libRaw->imgdata.params.use_camera_matrix = 1;
-    libRaw->imgdata.params.user_flip = 0;
-    libRaw->imgdata.params.use_fuji_rotate = 0;
-    libRaw->imgdata.params.gamm[0] = libRaw->imgdata.params.gamm[1] = 1;
-    libRaw->imgdata.params.highlight = 1;
-    libRaw->imgdata.params.output_bps = 16;
-    libRaw->imgdata.params.output_color = 4; // Kodak ProPhoto RGB D65
-    // libRaw->imgdata.params.output_tiff = 1;
-    
-    jclass cls = env->GetObjectClass(obj);
-    jmethodID mid = env->GetMethodID(cls, "progress", "(III)I");
-    callback data = {env, obj, mid};
-    if (mid != nullptr)
-        libRaw->set_progress_handler(my_progress_callback, (void *) &data);
-
-    status = libRaw->dcraw_process();
-    if (status != LIBRAW_SUCCESS)
-        return nullptr;
-
-    libraw_processed_image_t *image = libRaw->dcraw_make_mem_image(&status);
-    if (image == nullptr)
-        return nullptr;
-
-    constexpr int bands = 3;
-    const int buffer_size = bands * image->width * image->height;
-    jshortArray jimage_data = env->NewShortArray((jsize) buffer_size);
-    if (jimage_data != nullptr) {
-        env->SetShortArrayRegion(jimage_data, 0, buffer_size, (jshort *) image->data);
-    }
-    free(image);
-    return jimage_data;
+JNIEXPORT jshortArray JNICALL
+Java_com_lightcrafts_image_libs_LibRaw_unpackImage(JNIEnv *env, jobject obj,
+                                                   jboolean interpolate,
+                                                   jboolean half_size) {
+  jclass libRawClass = env->GetObjectClass(obj);
+  auto *libRaw = (LZLibRaw *)env->GetLongField(
+      obj, env->GetFieldID(libRawClass, "libRawObject", "J"));
+  if (libRaw == nullptr)
+    return nullptr;
+
+  int status = libRaw->unpack();
+  if (status != LIBRAW_SUCCESS)
+    return nullptr;
+
+  // params
+  if (half_size)
+    libRaw->imgdata.params.half_size = 1;  // -h
+  else if (interpolate)
+    libRaw->imgdata.params.four_color_rgb = 1;  // -f
+  else
+    libRaw->set_interpolate_bayer_handler(interpolate_bayer);
+
+  libRaw->imgdata.params.use_camera_wb = 1;  // -w
+//  libRaw->imgdata.params.use_camera_matrix = 1;  // -M
+
+  libRaw->imgdata.params.use_fuji_rotate = 0;  // -j
+  libRaw->imgdata.params.highlight = 1;  // -H 1
+  libRaw->imgdata.params.user_flip = 0;  // -t 0
+  libRaw->imgdata.params.gamm[0] = libRaw->imgdata.params.gamm[1] = 1;  // -g 1 1
+  libRaw->imgdata.params.no_auto_bright = 1;  // -W
+  libRaw->imgdata.params.output_bps = 16;  // -6
+  libRaw->imgdata.params.output_color = 4; // -o 4 (Kodak ProPhoto RGB D65)
```

Comment:
Looking at the Showfoto program, "Import Raw Using Libraw", they also have the D65 option, but it did not return as good results as the camera option.  They also have Automatic and Manual as choices.

Reply:
> I'm thinking that replacing 4 with the number 0 might do the trick

That is it!  Thank you very much!
![EOSR5hSLI000050NR0_lzn](https://github.com/ktgw0316/LightZone/assets/3116763/e76d5083-d3de-41e9-bc1a-0258ba4f91e0)

Since LZ handles camera2rgb matrix by itself in [RawAdjustmentsOperation](https://github.com/ktgw0316/LightZone/blob/a7b8acc1e6ffabc5dbf4bfbdcd047a7952ecc3a1/lightcrafts/src/com/lightcrafts/model/ImageEditor/RawAdjustmentsOperation.java#L136), I shouldn't use the `output_color` other than 0 (raw).